gridfs/grid_file.py

```diff
@@ -396,9 +396,10 @@ def __enter__(self) -> "GridIn":
     def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> Any:
         """Support for the context manager protocol.
 
-        Close the file and allow exceptions to propagate.
+        Close the file if no exceptions occur and allow exceptions to propagate.
         """
-        self.close()
+        if exc_type is None:
+            self.close()
```

Comment:
I think we should still mark the GridIn object as closed. Otherwise calling close explicitly after the with-statement would end up saving the file incorrectly:
```python
with GridIn(...) as file:
    file.write(...)
    raise Exception()
file.close() # This should be a no-op because the file is already marked as "closed"
```

Can you update the test to cover this case as well?